### DISCUSSION TITLE
Resolving function calls

### DISCUSSION THREAD

**Discussion by JEENB** [github/codeql#16254] [General]
I wrote a single hop caller callee query to generate call graph. 

```ql
from DataFlow::InvokeNode call, Function caller, Function callee, File caller_in, File callee_in, string func_name, string callee_name
where caller 	= call.getEnclosingExpr().getEnclosingFunction()
and callee 		= call.getACallee()
and func_name 	= caller.getName()
and callee_name = callee.getName()
and caller_in 	= caller.getFile()
and callee_in 	= callee.getFile()
select caller,  func_name,  callee, callee_name, caller_in, callee_in 
```

I realized it does not include all function calls. For example when I constructed the call graph for a file it had incomplete calls. For example when I run the query to generate all calls for handler, it does not return calls to handleCover or downloadFromUrl function. It only gives calls to isWarnmingCall, parse, checkLogin. Could there be a reason why this is the case?
The file I was talking is below. 

```js
import path from "path";
import uuid from "uuid/v4";

import checkLogin from "../../util/checkLoginToken";
import { handleCover, HandleCoverResult } from "../../util/handleCover";
import corsResponse from "../../util/corsResponse";

import downloadFromUrl from "../../util/downloadFromUrl";
import { isWarmingCall } from "../../util/isWarmingCall";

export const handler = async event => {
  if (isWarmingCall(event)) {
    return corsResponse({ coldStartPrevented: true });
  }

  const { userId, loginToken, url, similarBookCover = false } = JSON.parse(event.body);

  if (!(await checkLogin(userId, loginToken))) {
    return corsResponse({});
  }
  const { body, error } = await downloadFromUrl(url);

  if (error) {
    console.log("Error downloading", error);
    return corsResponse({ error: true });
  }
  const extension = path.extname(url) || ".jpg";

  let allResults: HandleCoverResult[];
  if (similarBookCover) {
    allResults = await Promise.all([void 0, handleCover(body, "small", `similar-books/${uuid()}${extension}`), void 0]);
  } else {
    const filePath = `${userId}/${uuid()}${extension}`;

    allResults = await Promise.all([
      handleCover(body, "mobile", `mobile-covers/${filePath}`),
      handleCover(body, "small", `small-covers/${filePath}`),
      handleCover(body, "medium", `medium-covers/${filePath}`)
    ]);
  }

  const success = allResults.filter(x => x).every(r => r.STATUS === "success");
  const [mobile, small, medium] = allResults;

  return corsResponse({
    success,
    mobile,
    small,
    medium
  });
};
```

**Comment by jketema** [comment]
Hi @JEENB,

Thanks for your question. `getACallee()` assumes the function being called and the module it is from are actually in your database. I cannot tell from your example whether this is the case for `downloadFromUrl` function and the `../../util/downloadFromUrl` module. Are you sure all the relevant modules and functions are actually part of your database?

**Reply by JEENB** [reply]
I create the database from the root directory so,  I'm assuming the module should be in the database. I don't know why this is the case though. Any leads would be helpful. This is what I used to generate the database
```codeql database create {db_path} -l {language} --source-root={full_path}```

**Reply by jketema** [reply]
Hi @JEENB ,

Apologies, I missed your answer. Did you try using the CodeQL VScode plugin to see what the AST of your code looks like? That might give you some insight whether there are actually calls in the locations you expect. It might be the case that `handleCover` is not actually call, because it's part of a list the AST will likely provide an answer to this question.
